File: C_version/Ant_colony.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <limits>
#include <random>
#include <cmath>

#include "ant.h"
#include "environment.h"

void display_progress_and_results(int current_iteration, int total_iterations, double best_distance, double alpha, double beta, double rho) {
    std::cout << "\r" << std::flush;

    double percentage = 100.0 * current_iteration / total_iterations;

    int bar_width = 50;
    int pos = bar_width * (current_iteration / static_cast<double>(total_iterations));
    std::cout << "[" << std::string(pos, '=') << ">" << std::string(bar_width - pos - 1, ' ') << "] " << std::fixed << std::setprecision(2) << percentage << "%";

    std::cout << " Best Distance: " << best_distance << " (α=" << alpha << ", β=" << beta << ", ρ=" << rho << ")" << std::flush;

    if (current_iteration == total_iterations) {
        std::cout << "[" << std::string(pos - 1, '=') << std::string(bar_width - pos - 1, ' ') << "] " << std::fixed << std::setprecision(2) << percentage << "%";
        std::cout << std::endl; // Move to the next line at the end of the process
    }
}

class AntColony {
private:
    int ant_population;
    int iterations;
    double alpha, beta, rho;
    Environment environment;
    std::vector<Ant> ants;

public:
    AntColony(int _ant_population, int _iterations, double _alpha, double _beta, double _rho)
        : ant_population(_ant_population), iterations(_iterations), alpha(_alpha), beta(_beta), rho(_rho), 
          environment(_rho, _ant_population) {
        // Initialize ants with random start locations
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(1, environment.n);

        // initialize the ants vector to the correct size
        ants.reserve(ant_population);
        
        // print the number of cities

        for (int i = 0; i < ant_population; ++i) {
            Ant newAnt = Ant(alpha, beta, dis(gen));
            newAnt.join(&environment);
            ants.emplace_back(newAnt);
        }
    }

    std::pair<std::vector<int>, double> solve() {
        std::vector<int> best_solution;
        double best_distance = std::numeric_limits<double>::infinity();

        for (int i = 0; i < iterations; ++i) {
            for (auto& ant : ants) {
                ant.run();
                if (ant.travelled_distance < best_distance) {
                    best_distance = ant.travelled_distance;
                    best_solution = ant.visited_locations;
                }
            }
            environment.update_pheromone_map(ants);
        }
        return {best_solution, best_distance};
    }
};


int main() {
    const int ants = 48, iterations = 50, iterations_per_level = 10;
    const std::vector<double> alphas = {0.8,0.9,1,1.1,1.2,1.3,1.4,1.5,1.6,1.7,1.8,1.9,2};
    const std::vector<double> betas = {1.5,1.6,1.7,1.8,1.9,2,2.1,2.2,2.3,2.4,2.5,2.6,2.7,2.8,2.9,3};
    const std::vector<double> rhos = {0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9};

    double best_distance = std::numeric_limits<double>::infinity();
    std::vector<int> best_solution;
    double alpha_best, beta_best, rho_best;

    for (auto alpha : alphas) {
        for (auto beta : betas) {
            for (auto rho : rhos) {
                for (int i = 1; i <= iterations_per_level; ++i) {
                    AntColony colony(ants, iterations, alpha, beta, rho);
                    auto [solution, distance] = colony.solve();

                    if (distance < best_distance) {
                        best_distance = distance;
                        best_solution = solution;
                        alpha_best = alpha;
                        beta_best = beta;
                        rho_best = rho;
                    }
                    display_progress_and_results(i,iterations_per_level,best_distance,alpha_best,beta_best,rho_best);
                }
            }
        }
    }

    std::cout << "Best settings: Alpha = " << alpha_best << ", Beta = " << beta_best << ", Rho = " << rho_best << std::endl;
    std::cout << "Best solution distance: " << best_distance << std::endl;
    std::cout << "Solution path: ";
    for (auto city : best_solution) {
        std::cout << city << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: Py_version/ant-colony.py
import numpy as np
import random
from environment import Environment
from ant import Ant 

# Class representing the ant colony
"""
    ant_population: the number of ants in the ant colony
    iterations: the number of iterations 
    alpha: a parameter controlling the influence of the amount of pheromone during ants' path selection process
    beta: a parameter controlling the influence of the distance to the next node during ants' path selection process
    rho: pheromone evaporation rate
"""
class AntColony:
    def __init__(self, ant_population: int, iterations: int, alpha: float, beta: float, rho: float):
        self.ant_population = ant_population
        self.iterations = iterations
        self.alpha = alpha
        self.beta = beta
        self.rho = rho 

        # Initialize the environment of the ant colony
        self.environment = Environment(self.rho, self.ant_population)

        # Initilize the list of ants of the ant colony
        self.ants = []

        # Initialize the ants of the ant colony
        for i in range(ant_population):
            # Initialize an ant on a random initial location 
            ant = Ant(self.alpha, self.beta, random.choice(range(self.environment.n)))
            ant.join(self.environment)
            self.ants.append(ant)

    # Solve the ant colony optimization problem  
    def solve(self):
        # The solution will be a list of the visited cities
        solution = []
        # Initially, the shortest distance is set to infinite
        shortest_distance = np.inf
        # Iterate over the number of iterations
        for i in range(self.iterations):
            # Run all the ants of the ant colony
            for ant in self.ants:
                ant.run()
                # Check if the ant found a better solution
                if ant.travelled_distance < shortest_distance:
                    shortest_distance = ant.travelled_distance
                    solution = ant.visited_locations
            # Update the pheromone map of the environment
            self.environment.update_pheromone_map(self.ants)
        return solution, shortest_distance


def main():
    random.seed(1337)
    # ant colony configuration
    ants = 48
    iterations_per_level = 50
    iterations = 100

    alphas = [0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5]
    betas = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6]
    rhos = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]

    results = {
        "Best": {
            "Alpha": None,
            "Beta": None,
            "Rho": None,
            "Solution": None,
            "Distance": None
        }    
    }

    best_distance = np.inf
    for alpha in alphas:
        for beta in betas:
            for rho in rhos:
                print("Alpha: ", alpha)
                print("Beta: ", beta)
                print("Rho: ", rho)
                for i in range(iterations_per_level):
                    ant_colony = AntColony(ants, iterations, alpha, beta, rho)
                    # Solve the ant colony optimization problem
                    solution, distance = ant_colony.solve()
                    if distance < best_distance:
                        best_distance = distance
                        results["Best"]["Alpha"] = alpha
                        results["Best"]["Beta"] = beta
                        results["Best"]["Rho"] = rho
                        results["Best"]["Solution"] = solution
                        results["Best"]["Distance"] = distance
                        print("Best solution: ", solution)
                        print("Best distance: ", distance)
            

    print(results)

if __name__ == '__main__':
    main()    